### ISSUE TITLE
Modify B_batch function to use its extend option

### ISSUE THREAD

**Issue by Flying-ow** [KindXiaoming/pykan#441]
In the B_batch function, there is a extend option, but you did not use it. Without grid extension, other functions in pykan package will have issues to get B-splines. The following is my modifications to B_batch function.

#----------------------------------Modified B_batch Function------------------------------------
import torch
def B_batch(x, grid, k=0, extend=True, device='cpu'):
    if extend:
        h = (grid[0, -1] - grid[0, 0]) / (grid.shape[1] - 1)
        for _ in range(1, k + 1):
            grid = torch.cat([grid[:, [0]] - h, grid], dim=1)
            grid = torch.cat([grid, grid[:, [-1]] + h], dim=1)
    
    x = x.unsqueeze(dim=2)
    grid = grid.unsqueeze(dim=0)
    
    if k == 0:
        value = (x >= grid[:, :, :-1]) * (x < grid[:, :, 1:])
    else:
        B_km1 = B_batch(x[:,:,0], grid=grid[0], k=k - 1, extend=False)
        value = (x - grid[:, :, :-(k + 1)]) / (grid[:, :, k:-1] - grid[:, :, :-(k + 1)]) * B_km1[:, :, :-1] + (
                    grid[:, :, k + 1:] - x) / (grid[:, :, k + 1:] - grid[:, :, 1:(-k)]) * B_km1[:, :, 1:]
    
    # In case grid is degenerate
    value = torch.nan_to_num(value)
    return value
#-------------------------------------------The End-----------------------------------------------
